### DISCUSSION TITLE
Kubernetes Pods fail silently when container runs out of memory

### DISCUSSION THREAD

**Discussion by Sinsin1367** [apache/airflow#24557] [General]
**Description**

In Airflow when a container runs out of memory the pod fails and OOMKilled is not published as a pod event. So nothing would show up in the logs. The OOMKilled is displayed as container status. We need to log container statuses if the pod fails to show in the logs that the container ran out and needs more resources to succeed. 

**Use case / motivation**

**Related Issues**

None

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by astolle** [comment]
Hi,

we have the same issue with Airflow v2.2.5, but when using the Kubernetes Executor.

When a POD get's killed by K8S because of OOM, Airflow logs "FAILED" in the scheduler log and nothing to the DAG logs. 

```
2022-06-17 06:22:38,609 kubernetes_executor.py INFO - Event: mypod.73f04e8e8d8142c084e0584a9ec7f9dc had an event of type MODIFIED
2022-06-17 06:22:38,611 kubernetes_executor.py ERROR - Event: mypod.73f04e8e8d8142c084e0584a9ec7f9dc Failed
```

Wished behaviour: DAG developers get the reason "why" the DAG (executor pod) failed. 

Best, Alex

**Comment by potiuk** [comment]
This is not how it works (and it will never work like that). It's mixing the responsibilities of Airflow with Deployment @astolle 

Only K8S knows why the POD has been killed and when the POD is killed because of OOM, it has no chance to even know and react on it. Making sure that resources are well allocated and monitoring it, is not an Airflow "task".  It's the job of monitoring deployment - you should monitor all OOMs and react to it. K8S is the "monitor" of the pods in this case not Airflow. 

Just make sure, monitoring of the K8S cluster is in place, the same way as you'd monitor any other application on K8S. Airflow is no different there @astolle @Sinsin1367. 

I am turning it into a discussion, but I believe (unless we change Airflow into resource management platform as well) that this something we will never change and it should be handled by proper cluster monitoring.

**Comment by 0robustus1** [comment]
I don't think that this should just be viewed as an OOM kill issue.

Ultimately Airflow schedules Pods into the system and wants to know whether the tasks it runs in these Pods (DAGs) succeed or fail. If they fail it should aim to provide a reasonable amount of context to allow a user to start their debugging.
Luckily Kubernetes does provide certain fields on the `status` subresource of a Pod to allow tools like Airflow to do just that.

I would suggest: Whenever a DAG fails (and Airflow reports that) it should provide context (like `status.reason`, `status.message` and `status.containerStatuses[*].state`.

**Reply by potiuk** [reply]
If you think that it solves the problem of your and can check it and provide a Pull Request fixing it - this is fantastic. Feel free to make a PR @0robustus1 . 

Airflow has more than 2000 contributors - most of them users like you, who missed a feature and wanted to add it. And you seem to know what you are doing and how to solve it, moreover you also have the envirnment to run all the tests and you could even deploy the fix localy in production or staging to see if it actually wokrs and test all those intrinsic complexities of how OOM condition might happen and propagate through the system, so you are one of the best people to contribute it :). The K8S executor code has been rewritten rather recently and it should be easy to find where status code is available, you can modify it, and by observing issues you had you might see if this is fixed - and PR when you find it works. 

It would be a great contribution if you can verify that it works and it can be done in such simple way as you mentioned and you are most welcome to do it!

However your problem **might** be different. I do not know if you correctly diagnosed it but OOM handling in K8S is far more complex than what you described - the "status.reason" and "status.message" is not enough and in many cases you might not get the information in status. But I might be wrong about it and it is indeed as simple as you said - then it will be also simple for you to test an implement - it's actually the only way to check which kind of OOM failures you experience.

Here Is a nice write-up about how it behaves and there are an interesting subtleties there - for example you might be surprised that OOM does not kill POD or container  but it kills random processes inside the pod if there is more than one process, and quite often status of the POD does not tell you that it eventually became killed by the OOM. 

https://mihai-albert.com/2022/02/13/out-of-memory-oom-in-kubernetes-part-4-pod-evictions-oom-scenarios-and-flows-leading-to-them/

Note that it is not k8s that decides to run OOM, this is done by the linux kernel, and K8S simply also just watches the events and reacts - sometimes poorly and in an unfair way. So the level of "decision" and status reporting and number of scenarios that the OOM event can propagate through k8S to particular Airflow task is quite high. So what you might be experiencing (and only you can tell that) this scenario where we are not able to retrieve the right information. I 'd say if we can display relevant information and it is available - we already do it, but maybe you prove me wrong. I'd love that to happen.

I'd say thought that it's quite likely we will not be able to handle OOM situation in the way that we always know this was an OOM  and are able to propagate that information to the logs of task. IMHO, t's much easier for you to monitor, diagnose and fix your OOM events in K8S.  You should not have OOMs anyway and having them is a sign of not Airflow having problem but simply misconfiguration of your deployment, or bad judgment of resource limitation your tasks are allocated (or maybe you do not apply limits at all). You should likely limit your resources better. 

I personally think it should be the job of your "deployment admin" to monitor and react to  the OOMs and investigate reasons for them. Especially that usually they might also be sign of cross-interference between running tasks and pods - for example you might get an OOM because ANOTHER task on that machine took all available memory, so having an OOM indication in the task might be  a poor advice sometimes and you need to be able to look at your "cluster" not at your single task.

But maybe we indeed could handle most of the regular OOMs be adding more information? Who knows ? Feel free to test it and make a PR.
